plugins/inputs/xtremio/xtremio.go

```diff
@@ -0,0 +1,324 @@
+	Cookie     *http.Cookie
+	tls.ClientConfig
+	Log telegraf.Logger
+}
+
+const sampleConfig = `
+  ## XtremIO Username
+  username = "user1" # required
+  ## XtremIO Password
+  password = "pass123" # required
+  ## XtremIO User Interface Endpoint
+  url = "https://xtremio.example.com/" # required
+  ## Metrics to collect from the XtremIO
+  collectors = ["bbus","clusters","ssds","volumes","xms"]
+
+  ## Optional TLS Config
+  # tls_ca = "/etc/telegraf/ca.pem"
+  # tls_cert = "/etc/telegraf/cert.pem"
+  # tls_key = "/etc/telegraf/key.pem"
+  ## Use SSL but skip chain & host verification
+  # insecure_skip_verify = false
+`
+
+// Description will appear directly above the plugin definition in the config file
+func (xio *XtremIO) Description() string {
+	return `Gathers Metrics From a Dell EMC XtremIO Storage Array's V3 API`
+}
+
+// SampleConfig will populate the sample configuration portion of the plugin's configuration
+func (xio *XtremIO) SampleConfig() string {
+	return sampleConfig
+}
+
+func (xio *XtremIO) Init() error {
+	if xio.Username == "" {
+		return fmt.Errorf("Username cannot be empty")
+	}
+	if xio.Password == "" {
+		return fmt.Errorf("Password cannot be empty")
+	}
+	if xio.URL == "" {
+		return fmt.Errorf("URL cannot be empty")
+	}
+
+	return nil
+}
+
+func (xio *XtremIO) Gather(acc telegraf.Accumulator) error {
+	tlsCfg, err := xio.ClientConfig.TLSConfig()
+	if err != nil {
+		return err
+	}
+	http.DefaultTransport.(*http.Transport).TLSClientConfig = tlsCfg
+	err = xio.Authenticate()
+	if err != nil {
+		return err
+	}
+	if xio.Cookie == nil {
+		return fmt.Errorf("no authentication cookie set")
+	}
+
+	availableCollectors := []string{"bbus", "clusters", "ssds", "volumes", "xms"}
+	var wg sync.WaitGroup
+	for _, collector := range xio.Collectors {
+		wg.Add(1)
+		go func(collector string) {
+			defer wg.Done()
+			if !contains(availableCollectors, collector) {
+				acc.AddError(fmt.Errorf("Specified Collector Isn't Supported: " + collector))
+				return
+			}
+
+			resp, err := xio.Call(collector)
+			if err != nil {
+				acc.AddError(err)
+				return
+			}
+			var arr []gjson.Result
+			if collector == "xms" {
+				arr = gjson.Get(resp, "xmss").Array()
+			} else {
+				arr = gjson.Get(resp, collector).Array()
+			}
+
+			for _, item := range arr {
+				itemSplit := strings.Split(gjson.Get(item.Raw, "href").Str, "/")
+				url := ""
+				if len(itemSplit) > 0 {
+					url = collector + "/" + itemSplit[len(itemSplit)-1]
+				} else {
+					continue
+				}
+
+				if collector == "bbus" {
+					wg.Add(1)
+					go xio.GatherBBUs(&wg, url, acc)
+				}
+				if collector == "clusters" {
+					wg.Add(1)
+					go xio.GatherClusters(&wg, url, acc)
+				}
+				if collector == "ssds" {
+					wg.Add(1)
+					go xio.GatherSSDs(&wg, url, acc)
+				}
+				if collector == "volumes" {
+					wg.Add(1)
+					go xio.GatherVolumes(&wg, url, acc)
+				}
+				if collector == "xms" {
+					wg.Add(1)
+					go xio.GatherXMS(&wg, url, acc)
+				}
```

Comment:
Please use a `switch` statement here. Furthermore, why do you add _another_ parallelization here? This will provide _no_ gain and only complicated things as you are already running in a go-routine.

Reply:
The first go-routine is for each collection type so we can run collectors in parallel. If you look at the code before this block, the first thing it does is retrieve all the objects under that collector and create the unique URLs for each object. This specific block of code lives in a for loop where it then sends each unique URL to a function to process it fully. 

So for something like SSDs, it's probably common for someone to have something like 50 SSDs in their storage array, and the idea is that it would be better to send 50 requests in parallel, rather than sequentially.

I can certainly pull that part out if you wish.

Reply:
Added comments for nested goroutines.